freqtrade/data/dataprovider.py

```diff
@@ -63,9 +79,100 @@ def _set_cached_df(
         :param dataframe: analyzed dataframe
         :param candle_type: Any of the enum CandleType (must match trading mode!)
         """
-        self.__cached_pairs[(pair, timeframe, candle_type)] = (
+        pair_key = (pair, timeframe, candle_type)
+        self.__cached_pairs[pair_key] = (
             dataframe, datetime.now(timezone.utc))
 
+    def add_external_df(
+        self,
+        pair: str,
+        timeframe: str,
+        dataframe: DataFrame,
+        candle_type: CandleType,
+    ) -> None:
+        """
+        Add the pair data to this class from an external source.
+
+        :param pair: pair to get the data for
+        :param timeframe: Timeframe to get data for
+        :param candle_type: Any of the enum CandleType (must match trading mode!)
+        """
+        pair_key = (pair, timeframe, candle_type)
+
+        # For multiple leaders, if the data already exists, we'd merge
+        self.__external_pairs_df[pair_key] = (dataframe, datetime.now(timezone.utc))
```

Comment:
I think for multiple leaders/producers, you'd track the data from each leader separately. Something like:
```
self.__external_producer_pairs_df[producer_key][pair_key] = (dataframe, datetime.now(timezone.utc))
```